### ISSUE TITLE
Why there is no DepthwiseConv1D function in TensorFlow?

### ISSUE THREAD

**Issue by yjiangling** [tensorflow/tensorflow#48557]
The function DepthwiseConv2D、SeparableConv1D and SeparableConv2D are all supported in TensorFlow, however, the function DepthwiseConv1D is still not supported yet. But in fact, the function of SeparableConv1D is conduct by a sequence of a DepthwiseConv and a PointWiseConv, so it seems that it's very convient to add DepthwiseConv1D function to the TensorFlow.

**Comment by fchollet** [comment]
Sure, that seems reasonable. Would you like to open a PR?

**Comment by around-star** [comment]
@fchollet I can work on it.

**Comment by nikitamaia** [comment]
@around-star assigning this issue to you. Thanks!

**Comment by AdityaKane2001** [comment]
@around-star 
Would like to work on this with you.

**Comment by bhack** [comment]
I think that we could assign assign two contributors but you need to collaborate on the same external fork cause we expect a single PR. Nikita what do you think?

**Comment by andreselizondo-adestech** [comment]
I would like to request support for `causal` padding for the new DepthwiseConv1D layer.
Thanks :)

**Comment by AdityaKane2001** [comment]
@bhack 
Myself and @around-star have some code ready. Unfortunately, we are not able to build it multiple times due to limited resources. Thus, it is has become quite difficult to debug. Can you suggest some solution?

**Comment by bhack** [comment]
We are working on this at https://github.com/tensorflow/tensorflow/pull/48421 

In the meantime If you can complete the first build the with this param it will easier with the next builds if you don't resync too much you branch with master: https://docs.bazel.build/versions/master/remote-caching.html#disk-cache

**Comment by bhack** [comment]
@AdityaKane2001 Are you able to complete the first build with your local machine?

**Comment by AdityaKane2001** [comment]
@bhack 
I was not able to complete the build, I am getting following error message:
`ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted: Repository command failed
INFO: Could not find files for the given pattern(s).`

**Comment by bhack** [comment]
It was fixed at https://github.com/microsoft/vcpkg/pull/16315

**Comment by AdityaKane2001** [comment]
@bhack 
I updated python to 3.9.2 and reinstalled the MS Build Tools but the error still persists.

**Comment by bhack** [comment]
It was closed as stale on our side  https://github.com/tensorflow/tensorflow/issues/46902 /cc @mihaimaruseac 

In the meantime can you build with out Docker devel image at https://www.tensorflow.org/install/source#docker_linux_builds?

**Comment by AdityaKane2001** [comment]
Sure, I'll give it a try. Will I be able to use the bazel cache somehow? The previous build had compiled around 11k objects.

**Comment by bhack** [comment]
No it will be invalidated as the platform is totally different. Are you sure that you are compiling on updated Build tools with python 3.9.2 cause the `copysign` MACRO was removed in Python 3.9.2 so it seems strange that you still have these conflicts.

**Comment by AdityaKane2001** [comment]
Yes, it seemed a bit weird to me too. 
I have installed docker and pulled the container, I'll try to build.

**Comment by AdityaKane2001** [comment]
@bhack 
I tried to build using docker. It was going fine for a while, and then it got stuck at
` Compiling tensorflow/core/kernels/bias_op.cc [for host]; 2086s local, remote-cache` for 30 mins or so
I cancelled the build. When I restarted it, it started doing all over again from scratch.

**Comment by bhack** [comment]
Have you compiled with --disk_cache= ?

**Comment by AdityaKane2001** [comment]
Yes I did

**Comment by bhack** [comment]
> Yes I did
> 

Have you restarted the container?

**Comment by AdityaKane2001** [comment]
```cmd
bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package --disk_cache=../docker_build_cache --local_cpu_resources=4 
 --local_ram_resources=HOST_RAM*.5
```
was the command I ran

**Comment by AdityaKane2001** [comment]
>Have you restarted the container?


Not yet

**Comment by bhack** [comment]
Remember that the containers are ephimeral. You need to pass the disk cache as volume so It will be persistent across containers destroy/creation instances.

**Comment by AdityaKane2001** [comment]
> You need to pass the disk cache as volume so It will be persistent across containers destroy/creation instances.

Could you please tell a bit more about that?

**Comment by bhack** [comment]
Read https://docs.bazel.build/versions/master/bazel-container.html#build-abseil-project-from-your-host-machine-with-directory-mounting

And you need to do the swme also for the volume/path that you want to use for the `--disk_cache`

**Comment by AdityaKane2001** [comment]
Thanks a lot, I'll take a look at it.

**Comment by AdityaKane2001** [comment]
@bhack 
Please give me some time to do this. I am quite new to Docker, but I'll try my best. Thanks for your support and patience.

**Comment by bhack** [comment]
Ok you have two approaches in that Doc: 
- with a Git checkout on the host
- with the checkout in the container  (in our image we also already have a Tensorflow checkout).

So as you like you but remember to set a volume mount also for `--disk_cache` directory.

**Comment by AdityaKane2001** [comment]
@bhack 
Facing the following error:
```
 /tensorflow_src/tensorflow/lite/python/BUILD:55:10 C++ compilation of rule '@llvm-project//llvm:Passes' failed (Exit 4): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 50 argument(s) skipped)
```
I used git checkout in the container, and my fork is about a week old, not more than that.
As per your instructions, I set up a volume and used it in the build commands. Following are the relevant commands that were executed:

```
> docker run -it -w /tensorflow_src -v tfcache:/mnt -e HOST_PERMS="$(id -u):$(id -g)" tensorflow/tensorflow:devel bash  //tfcache is the name of my volume

# git pull https://github.com/AdityaKane2001/tensorflow.git
# ./configure
#  bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package --disk_cache /mnt --local_ram_resources=HOST_RAM*.5 --local_cpu_resources=6

```

**Comment by bhack** [comment]
So you have used the first solution (to use the checkout inside the cotaienr).
But with the commands I see are you sure that you are building in the right folder?

I see that you have `-w /tensorflow_src` so your already inside the tensorflow master `git checkout` that is available in the image.

If you want to work with your remote in this  checkout directory I suggest you to just [add a new remote](https://docs.github.com/en/github/getting-started-with-github/managing-remote-repositories#adding-a-remote-repository) with your fork and not to do a complete checkout in that directory.

**Comment by AdityaKane2001** [comment]
> a complete checkout in that directory.


Do you mean I should pull my fork in a new directory and work in that directory?

**Comment by bhack** [comment]
As you wan. You have 3 main solutions:

- You could mount your checkout from your host with an extra `-v`
- You can add your remote in /tensorflow_src
- You could checkout in new  different directory ephimeral directory.

All these will work with its own pro and cons.

**Comment by AdityaKane2001** [comment]
I I use this method, each time I use `docker run`  with the volumes attached, I'll be ready to go, right? Because right now I committed the container to a new local image. Yet, I'm not sure whether I will be able to access the changes made during the session.

**Comment by bhack** [comment]
Yes, in that case you could use a volume or a local host path.

**Comment by AdityaKane2001** [comment]
Ok. Could you please tell how can I resolve the error?

```
 /tensorflow_src/tensorflow/lite/python/BUILD:55:10 C++ compilation of rule '@llvm-project//llvm:Passes' failed (Exit 4): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 50 argument(s) skipped)```

**Comment by bhack** [comment]
I don't think that it is related to your commits, you could rebase or merge the current master on your branch.
Also you could try to compile only your target (this example is for convolutional you can change the string with your own new layer):

```
bazel bazel build --config=opt //tensorflow/python/keras/layers:convolutional --disk_cache /mnt --local_ram_resources=HOST_RAM*.5 --local_cpu_resources=6  
```

**Comment by AdityaKane2001** [comment]
Ok, so as you said earlier, I'll mount two volumes and try to build again.

**Comment by bhack** [comment]
Yes It was basically the first of the two solutions already mentioned at https://docs.bazel.build/versions/master/bazel-container.html#build-abseil-project-from-your-host-machine-with-directory-mounting

In that example was `-v /src/workspace:/src/workspace`.

If you still want to have the same access from your host remember to use 
`  -e USER="$(id -u)" \
  -u="$(id -u)" \` as in the example

**Comment by AdityaKane2001** [comment]
Got it 👍

**Comment by AdityaKane2001** [comment]
@bhack 
Used docker compose this time. As discussed, I used 2 volumes: 
```
#docker-compose.yaml

version: "3"

services: 
    tf:
        image: tensorflow/tensorflow:devel
        volumes: 
            - ./cache:/cache
            - ./repo:/repo
    

```
Used the following command:
```
bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package --disk_cache ../cache --local_ram_resources=HOST_RAM*.5 --local_cpu_resources=4
```

Encountered the following errors on two successive builds : 
```
ERROR: /repo/tensorflow/compiler/mlir/tensorflow/BUILD:411:15: C++ compilation of rule '//tensorflow/compiler/mlir/tensorflow:tensorflow_ops_n_z' failed (Exit 4): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 111 argument(s) skipped)
gcc: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /repo/tensorflow/lite/python/BUILD:55:10 C++ compilation of rule '//tensorflow/compiler/mlir/tensorflow:tensorflow_ops_n_z' failed (Exit 4): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 111 argument(s) skipped)
INFO: Elapsed time: 2011.170s, Critical Path: 631.06s
INFO: 1083 processes: 155 internal, 928 local.
```

and 

```
ERROR: /repo/tensorflow/core/kernels/BUILD:5287:18: C++ compilation of rule '//tensorflow/core/kernels:training_ops' failed (Exit 4): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 86 argument(s) skipped)
gcc: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /repo/tensorflow/lite/python/BUILD:55:10 C++ compilation of rule '//tensorflow/core/kernels:training_ops' failed (Exit 4): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 86 argument(s) skipped)
INFO: Elapsed time: 5235.750s, Critical Path: 444.55s
``` 

Could you please tell how can I resolve this error? 
Thanks

**Comment by bhack** [comment]
Have you rabased your branch?

**Comment by AdityaKane2001** [comment]
Yes

**Comment by bhack** [comment]
Probably you could try to reduce your Resources (especially RAM). See https://github.com/tensorflow/tensorflow/issues/35051#issuecomment-623751066

**Comment by AdityaKane2001** [comment]
I'll try

**Comment by bhack** [comment]
If you still have the same issue after tuning the resources please post a comment at https://github.com/tensorflow/tensorflow/issues/39947

**Comment by AdityaKane2001** [comment]
Sure

**Comment by AdityaKane2001** [comment]
Requesting to close, as it is now implemented in [#14863](https://github.com/keras-team/keras/pull/14863) of keras-team/keras.

**Comment by andreselizondo-adestech** [comment]
@AdityaKane2001 Would it be possible to support `causal` padding?
This is something I need for a streaming model.

**Comment by AdityaKane2001** [comment]
@andreselizondo-adestech 
Sorry, but I wasn't the one who contributed this code. Nor am I aware of its internal working. Please open an issue in keras-team/keras for this.

**Comment by bhack** [comment]
I'have any permission on this repository. @mihaimaruseac can you close this?

**Comment by AdityaKane2001** [comment]
@bhack 
Thanks for all the support. Learnt a lot in this one.
